### DISCUSSION TITLE
Is it possible to use functions in custom dll that return objects of the model?

### DISCUSSION THREAD

**Discussion by bernatagulloesbrina** [TabularEditor/TabularEditor3#432] [🙋♀️TE3 Q&A] [answered]
it is a very niche question, so I'll just leave it here in case someone has succeded. 

I'm creating custom functions in a custom dll to reuse from tabular editor scripts 
I've succeeded creating boolean functions (which can have out parameters, so it's not all lost) 
however when I tried creating a funciton that returns a table or calculated table I could not go around the error
"Object Reference Not Set to an instance of an object."

Model.AddCalculatedTable should return a table (or CalculatedTable) object right? 

either my lacking tomwrapper knowledge and/or c# knowledge are to blame for this question

this works
```
public static bool CreateMeasureTable
            (
                Model model, 
                string defaultTableName = "Some Measures"
              )
        {

            string tableExpression = "{0}";

            model.AddCalculatedTable(name: defaultTableName, expression: tableExpression);
            
            return true;

        }
```
this does not

```
        public static Table CreateMeasureTable
            (
                Model model,
                string defaultTableName = "Some Measures",
             )
        {

            string tableExpression = "{0}";

            return model.AddCalculatedTable(name: defaultTableName, expression: tableExpression);

            

        }

```
am I doing something wrong? 

I might stick with the boolean approach, but wondering if in some cases I could do the object approach

**Comment by otykier** [comment]
I think the error you're seeing originates from the outer script that calls into your custom dll. Could you share that script here as well?

**Comment by bernatagulloesbrina** [comment, marked as answer]
Sure, 

it's something like this. I tried with Table and CalculatedTable  

```
#r "C:\Users\ESAgulloBe\OneDrive - NESTLE\AAA Repo\ExtendedTOMWrapper\bin\x64\Debug\ExtendedTOMWrapper.dll"

using ExtendedTOMWrapper; 


Table t = Model.AddCalculatedTable(name: "myName2", expression: "{0}");
Output(t.Name);
```

**Reply by otykier** [reply]
But in that code, you're not using the `CreateMeasureTable` method from your dll?

**Reply by bernatagulloesbrina** [reply]
oups! apologies, here's the code

```
#r "C:\Users\ESAgulloBe\OneDrive - NESTLE\AAA Repo\ExtendedTOMWrapper\bin\x64\Debug\ExtendedTOMWrapper.dll"

using ExtendedTOMWrapper; 

string myName;
CalculatedTable myTable = TableUtils.CreateMeasureTable(defaultTableName:"myMeasureTable",model:Model);
```

**Reply by otykier** [reply]
Looks good to me (assuming `TableUtils` is the name of the class in which the `CreateMeasureTable` static method resides).

Maybe try to add some null checks here and there? For example, try to add something like:

```csharp
if(model == null) throw new ArgumentNullException("model");
```

just above the line where you call `model.AddCalculatedTable` in your dll code. You should then see a different error message if the model object does not get correctly passed to `CreateMeasureTable`.

**Reply by bernatagulloesbrina** [reply]
Well, I'm not sure what is it, but now I put the null check and it works! (so it was not passing a null, so it should have worked, but whatever)
Thank you for your patience and dedication!
